### PR TITLE
Shard scheduled mini blocks pre-validation

### PR DESCRIPTION
**Pull request by SebastianMarian**
* Implemented shard scheduled mini blocks pre-validation
* Added scheduled mini blocks in the proposed body

**How to be tested**

Start a normal system test with txs and sc calls
In epoch 5 scheduled mini blocks mechanism will be activated.
In epoch 6 we should check the log files and messages like: "createScheduledMiniBlocks has been started", "createScheduledMiniBlocks has been finished" and "scheduled mini blocks from me created" should be found

### CODE REVIEW

**Comment by gabi-vuls** [comment]
System test  **_passed_**

No messages like : " createScheduledMiniBlocks has been started" , " createScheduledMiniBlocks has been finished " , "scheduled mini blocks created" BEFORE epoch 6. 

Messages are present only after epoch 6. 
All these 3 type of messages are present.

File: process/block/preprocess/transactionsV2.go [hunk 1]
```diff
@@ -428,29 +470,183 @@ func (txs *transactions) createScheduledMiniBlocks(
...
+			log.Debug("scheduled mini block is not created", "shard", receiverShardID)
 			continue
 		}
 
-		log.Debug("this sc call could be added in scheduled mini blocks",
-			"nonce", tx.Nonce,
-			"hash", txHash,
-			"sender shard", senderShardID,
-			"receiver shard", receiverShardID)
+		numNewMiniBlocks := 0
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:494]: take some part out into a new function. this is still way too big.

File: process/block/preprocess/transactionsV2.go [hunk 2]
```diff
@@ -478,14 +678,64 @@ func (txs *transactions) displayProcessingResults(
...
+
+func getAdditionalTimeFunc() func() bool {
+	startTime := time.Now()
+	return func() bool {
+		return additionalTimeForCreatingScheduledMiniBlocks > time.Since(startTime)
+	}
+}
+
+// isScheduledMiniBlock returns if the given mini block is of type scheduled or not
+func isScheduledMiniBlock(miniBlock *block.MiniBlock) bool {
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:736]: you could implement this at the miniblock level as a public function.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: process/block/displayBlock.go [hunk 3]
```diff
@@ -165,8 +165,14 @@ func (txc *transactionCounter) displayTxBlockBody(lines []*display.LineData, bod
 	for i := 0; i < len(body.MiniBlocks); i++ {
 		miniBlock := body.MiniBlocks[i]
 
+		mbTypeStr := miniBlock.Type.String()
+		if isScheduledMiniBlock(miniBlock) {
+			//mbTypeStr = block.ScheduledBlock.String()
+			mbTypeStr = "ScheduledBlock"
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/displayBlock.go:171]: done

File: process/block/displayMetaBlock.go [hunk 4]
```diff
@@ -177,8 +177,14 @@ func (hc *headersCounter) displayTxBlockBody(lines []*display.LineData, body *bl
 	for i := 0; i < len(body.MiniBlocks); i++ {
 		miniBlock := body.MiniBlocks[i]
 
+		mbTypeStr := miniBlock.Type.String()
+		if isScheduledMiniBlock(miniBlock) {
+			//mbTypeStr = block.ScheduledBlock.String()
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/displayMetaBlock.go:182]: done

File: process/block/interceptedBlocks/interceptedMiniblock.go [hunk 5]
```diff
@@ -112,9 +112,10 @@ func (inMb *InterceptedMiniblock) integrity() error {
 		}
 	}
 
-	if len(miniblock.Reserved) > 0 {
-		return process.ErrReservedFieldNotSupportedYet
-	}
+	//TODO: Remove this condition after the final decision about reserved field from mini block
+	//if len(miniblock.Reserved) > 0 {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/interceptedBlocks/interceptedMiniblock.go:116]: I know, this name is used in almost all data structures: MiniBlockHeader, MiniBlock, Header, MetaBlock, etc

File: process/block/preprocess/transactions.go [hunk 6]
```diff
@@ -478,6 +505,31 @@ func (txs *transactions) processTxsFromMe(
 		return process.ErrTimeIsOut
 	}
 
+	if txs.flagScheduledMiniBlocks.IsSet() {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/preprocess/transactions.go:514]: done

File: process/block/preprocess/transactions.go [hunk 7]
```diff
@@ -780,6 +832,24 @@ func (txs *transactions) CreateAndProcessMiniBlocks(haveTime func() bool) (block
 		return make(block.MiniBlockSlice, 0), nil
 	}
 
+	if txs.flagScheduledMiniBlocks.IsSet() {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/preprocess/transactions.go:841]: done

File: process/block/preprocess/transactionsV2.go (hunk 1 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:506]: WIll be done in the last PR

File: process/transaction/shardProcess.go [hunk 8]
```diff
@@ -210,6 +210,20 @@ func (txProc *txProcessor) ProcessTransaction(tx *transaction.Transaction) (vmco
 	return vmcommon.UserError, txProc.executingFailedTransaction(tx, acntSnd, process.ErrWrongTransaction)
 }
 
+// VerifyTransaction verifies the account states in respect with the transaction data
+func (txProc *txProcessor) VerifyTransaction(tx *transaction.Transaction) error {
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/transaction/shardProcess.go:214]: done

File: process/block/preprocess/transactionsV2.go (hunk 2 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:749]: done

**Review by sasurobert — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**